src/mongocxx/test/collection.cpp

```diff
@@ -2808,4 +2812,141 @@ TEST_CASE("Ensure that the WriteConcernError 'errInfo' object is propagated", "[
     REQUIRE(contains_err_info);
 }
 
+TEST_CASE("expose writeErrors[].errInfo", "[collection]") {
+
+    // A helper for checking that an error document is well-formed according to our requirements:
+    auto writeErrors_well_formed = [](const bsoncxx::document::view& reply_view) -> bool {
+        if (!reply_view["writeErrors"]) {
+            return false;
+        }
+
+        const auto& errdoc = reply_view["writeErrors"][0];
+
+        auto error_code = errdoc["code"].get_int32();
+
+        // The code should always be 121 (DocumentValidationFailure):
+        if (121 != error_code) {
+            std::ostringstream os;
+            os << "writeErrors expected to have code 121, but had " << error_code << " instead";
+            throw std::runtime_error(os.str());
+        }
+
+        // We require the "details" field be present:
+        if (!errdoc["errInfo"]["details"]) {
+            throw std::runtime_error("no \"details\" field in \"writeErrors\"");
+        }
+
+        return true;
+    };
+
+    // Set up our test environment:
+    instance::current();
+
+    auto client_opts = mongocxx::options::client();
+
+    // Listen to the insertion-failed event: we want to get a copy of the server's
+    // response so that we can compare it to the thrown exception later:
+    apm_opts.on_command_succeeded(
+        [&writeErrors_well_formed](const mongocxx::events::command_succeeded_event& ev) {
+            if(!writeErrors_well_formed(ev.reply()))
+        });
+
+    client_opts.apm_opts(apm_opts);
+
+    auto mongodb_client = mongocxx::client(uri{}, client_opts);
+
+    database db = mongodb_client["prose_test_expose_details"];
+
+    const std::string collname{"mongo_cxx_driver-expose_details"};
+
+    // Drop the existing collection, if any:
+    db[collname].drop();
+
+    // Make a new collection with validation checking:
+    collection coll = db.create_collection(
+        collname,
+        make_document(kvp("validator",
+                          make_document(kvp("field_x", make_document(kvp("$type", "string")))))));
+
+    SECTION("insert a valid document") {
+        bsoncxx::builder::basic::document entry;
+
+        entry.append(kvp("_id", bsoncxx::oid()), kvp("field_x", "value0"));
+
+        REQUIRE_NOTHROW(coll.insert_one(entry.view()));
+    }
+
+    SECTION("cause a type violation on insert") {
+        bsoncxx::builder::basic::document entry;
+
+        entry.append(kvp("_id", bsoncxx::oid()), kvp("field_x", 42));
+
+        try {
+            coll.insert_one(entry.view());
+
+            // We should not make it here:
+            CHECK(false);
+        } catch (const operation_exception& e) {
+            auto rse = e.raw_server_error();
+
+            CHECK(rse.has_value());
+
+            CHECK(writeErrors_well_formed(*rse));
```

Comment:
It is an error if no exception is thrown. I suggest adding a boolean flag to assert an `operation_exception` is thrown.  Something like:
```cpp
bool operation_exception_thrown = false;

try {
    coll.insert_one(entry.view());

    // We should not make it here:
    CHECK(false);
} catch (const operation_exception& e) {
    exception_thrown = true;
    // ...
}

CHECK (operation_exception_thrown);
```

Reply:
My code above should catch this, under "We should not make it here". Catch2 provides some helpful macros, but unfortunately the version that we'd need in this particular case (since we have to do extra checking on the value) requires more manipulation than I see provided by the library. I /think/ the case is covered, though.

Reply:
Good point. Resolving.